torch/distributed/_shard/sharding_spec/api.py

```diff
@@ -110,3 +110,65 @@ def __post_init__(self):
             rank = len(shard.shard_offsets)
 
         validate_non_overlapping_shards_metadata(self.shards)
+
+
+def _infer_sharding_spec_from_shards_metadata(shards_metadata):
+    """
+    Infer the sharding spec from the metadata of each shard of a ShardedTensor.
+    If the tensor is sharded only on one dimension, we then assume it's chunk sharded.
+    Otherwise, we assume it's enum sharded.
+
+    Args:
+        shards_metadata (List[ShardMetadata]): List of Metadata of local shards.
+
+    Returns:
+        A :class:`torch.distributed._shard.sharding_spec.ShardingSpec` object of sharding
+            spec for one sharded tensor.
+    """
+    placements = []
+    chunk_sharding_dim = None
+    chunk_offset_list = []
+    shard_size_list = []
+    # collect local shard metadatas from the global sharded_tensor_metadata
+    for shard_metadata in shards_metadata:  # type: ignore[attr-defined]
+        placements.append(shard_metadata.placement)
+        local_offsets = shard_metadata.shard_offsets
+        chunk_offset_list.append(sum(local_offsets))
+        shard_size_list.append(shard_metadata.shard_sizes)
+        shard_dims = [idx for idx, e in enumerate(local_offsets) if e != 0]
+        if len(shard_dims) == 0:
+            continue
+        if len(shard_dims) != 1:
+            chunk_sharding_dim = None
+            break
+        if not chunk_sharding_dim:
+            chunk_sharding_dim = shard_dims[0]
+        elif chunk_sharding_dim != shard_dims[0]:
+            chunk_sharding_dim = None
+            break
+
+    if chunk_sharding_dim is not None:
+        # Ensure we infer the correct placement order from offsets
+        placements = [
+            x for _, x in sorted(zip(chunk_offset_list, placements), key=lambda e: e[0])
+        ]
+        chunk_spec = ChunkShardingSpec(
+            dim=chunk_sharding_dim,
+            placements=placements,
+        )
+        shard_sizes = [
+            x[chunk_sharding_dim]
+            for _, x in sorted(zip(chunk_offset_list, shard_size_list))
+        ]
+        if len(shard_sizes) == 1 or (
```

Comment:
can you also make a comment to explain what this logic used for?

Reply:
sure, this is for corner cases like when enum sharding will also shard by one dimension but the offset is not the same, aka, the tensor is not evenly distributed.